src/cargo/core/resolver/encode.rs

```diff
@@ -578,6 +596,10 @@ impl<'a> ser::Serialize for Resolve {
             root: None,
             metadata,
             patch,
+            version: match self.version() {
+                ResolveVersion::V3 => Some(1),
```

Comment:
Shouldn't it rather start at 3? Would be less confusing IMO

Reply:
Only for someone reading Cargo's source code, but for everyone else seeing changes to their `Cargo.lock` for the first time it's probably much less surprising to start at 1 instead of 3.

Reply:
The v2 format has changes to `Cargo.lock` as well, in fact much larger ones. People wouldn't be seeing it for the first time. But ultimately this question is bikeshedding so I won't continue discussing it. Your choice. shrug.

Reply:
If we are bikeshedding it, I'd go for 3. I'm imagining a docs or blog post explaining the history of the lockfile format, for example the one explaining the corner cases where this brakes things. 

> "So now we get to the third change to the format. It is called `version=1`... 
back in the first version (the one without a version specified not the one called "1")... 
in the new third format (the one called "1" not the original.)" 

We are making a complicated toppik harder to talk about. On the other hand if we start with 3, that blog post will have to explain what happened to 1 and 2. That feels like a strateforered one sentence explanation. 

>", which predates the version field."

Anyway defintly a bikeshed.

Reply:
It's absolutely a bikeshed, but I think on balance it'll make Cargo a little more impenetrable for new contributors if the version numbers don't match. I think I'd rather see an explanation *once* for why we're starting with 3 (and what happened to 2), rather than having a mismatch for all time.

Reply:
I'm getting back into this sub-discussion for this one comment only. The format has changed prior to the "lockfiles 2.0" change. Some examples:

* 5430db61f6c2668ec94945c91a1df22015ff1cf8 has put checksums into Cargo.lock... yes there used to be a time when there were no checksums :p.
* #4571 has removed `[root]`
* #6180 has added a `@generated` comment at the top
* #7070 added the new format commonly referred to as lockfiles v2

Not guaranteeing for completeness of this list. I'm not sure what the "perfect" version number should be. But if #7070 says that it introduced lockfiles v2, even if that was technically wrong, it's better to continue with 3 than jumping over version numbers or starting with 1 again.

I've also wondered about semver lockfile versions where the major version indicates compatibility breakage (like the one which this PR introduces) and the minor version indicates small changes that don't break older cargo's but cargo may not generate the lockfile 100% the same way, e.g. #6180 would be one such change. If cargo supports writing lockfiles with the major version but the minor version is unsupported, it should do a more thorough check than a strcmp and only update if needed (to prevent needless back and forth, and only have back and forth if there is a different reason to update the lockfile). Basically I propose getting control flow into the body of [this](https://github.com/rust-lang/cargo/blob/abc8463469111c5630f115dbf27f9fd6b9737ab7/src/cargo/ops/lockfile.rs#L170) if check if the minor version is unsupported but the major is. The reason why I haven't proposed this earlier was because I doubt that there could be minor version increases like the generated comment addition in the future. Maybe my imagination is limited, idk. The version can be made semver in the future as well, it'd require breaking support for old cargo, aka a major number increase, and mess up the nice error messages for old cargo (turn "this version isn't supported" into "expected number, got string"). It's really a bikeshed.

Reply:
I've just gone ahead and flagged this as `version = 3`